daemon/emer-aggregate-tally.c

```diff
@@ -259,50 +277,29 @@ emer_aggregate_tally_store_event (EmerAggregateTally  *self,
                                   gint64               monotonic_time_us,
                                   GError             **error)
 {
-  g_autofree gchar *tally_path = NULL;
-  g_autofree gchar *dirname = NULL;
+  const char *UPSERT_SQL =
+    "INSERT INTO tally (date, event_id, unix_user_id, aggregate_key, payload, counter) "
+    "VALUES (?, ?, ?, ?, ?, ?) "
+    "ON CONFLICT (date, event_id, unix_user_id, aggregate_key, payload) "
+    "DO UPDATE SET counter = tally.counter + excluded.counter;";
+
   g_autofree gchar *date = NULL;
-  g_autoptr(GVariant) timer_variant = NULL;
-  g_autoptr(GError) local_error = NULL;
-  guint32 previous_counter;
+  sqlite3_stmt *stmt = NULL;
 
   date = format_datetime_for_tally_type (datetime, tally_type);
-  tally_path = get_tally_path_from_aggregate_timer (self, unix_user_id,
-                                                    event_id, aggregate_key,
-                                                    date);
-
-  /* It might happen that the same application was opened and closed
-   * multiple times at the same day by the same user. In this case,
-   * we want to increase the day's counter instead of simply override
-   * it.
-   */
-  previous_counter = find_previous_aggregate_timer_counter (self, tally_path);
 
-  dirname = g_path_get_dirname (tally_path);
-  ensure_folder_exists (self, dirname, &local_error);
-  if (local_error)
-    {
-      g_propagate_error (error, g_steal_pointer (&local_error));
-      g_prefix_error (error, "Error saving aggregate timer to tally: ");
-      return FALSE;
-    }
+  CHECK (sqlite3_prepare_v2 (self->db, UPSERT_SQL, -1, &stmt, NULL));
 
-  timer_variant = aggregate_timer_to_variant (unix_user_id, event_id,
-                                              aggregate_key, payload,
-                                              counter + previous_counter,
-                                              monotonic_time_us);
+  CHECK (sqlite3_bind_text (stmt, 1, date, -1, SQLITE_TRANSIENT));
+  CHECK (sqlite3_bind_text (stmt, 2, g_variant_print (event_id, TRUE), -1, g_free));
+  CHECK (sqlite3_bind_int (stmt, 3, * ((int *) &unix_user_id)));
+  CHECK (sqlite3_bind_blob (stmt, 4, g_variant_print (aggregate_key, TRUE), -1, g_free));
```

Comment:
Why are you storing the printed variant as a blob? Why not `g_variant_get_data()`?